Trace SDK: Provide definitions for readable and read/write span. (#669)
open-telemetry/opentelemetry-specification: specification/trace/sdk.md

reference to their `TracerProvider` and access configuration only via this
reference.

## Additional Span Interfaces

The [API-level definition for Span's interface](api.md#span-operations)
only defines write-only access to the span.
This is good because instrumentations and applications are not meant to use the data
stored in a span for application logic.
However, the SDK needs to eventually read back the data in some locations.
Thus, the SDK specification defines two new terms:

* **Readable span**: A span interface satisfies the requirements of being a

>   readable span if it allows to retrieve all information

Reviewer: Should this be without overhead? I noticed that at least in Java, the only way to read all the information is to convert to `SpanData` which doesn't seem good for just reading. For example

https://github.com/open-telemetry/opentelemetry-java/blob/cb04118a21d86c7219cb2728d8b3cc1bb0eb3443/sdk_extensions/zpages/src/main/java/io/opentelemetry/sdk/extensions/zpages/TracezSpanBuckets.java#L49

Convert just to read status.

It does seem to give more importance in separating export from the other span processor methods

Author: I don't know if we should specify that here. I think it's a "quality of implementation" issue. Of course it would be nice if it works without overhead, but there may be other trade offs.

  that was added to the span (e.g. all attributes, events, (parent) `SpanContext`,
  etc.).
  

>   A readable span interface MUST provide a method called `getInstrumentationLibrary` (or similar),

Reviewer 2: Why is this special case than other? It also should probably have a getResource correct if the idea is to list all methods/functions that are not corresponding to Span API?

Author: Good question, I did not change that part in this PR. I agree that probably the resource should also be listed here, since they are both non-obvious IMHO.

Author: I commented on https://github.com/open-telemetry/opentelemetry-specification/issues/158#issuecomment-662324428 and un-linked that PR, so that it won't close this issue.

  that returns the `InstrumentationLibrary` information held by the `Tracer` that created the span.

> * **Read/write span**: A span interface satisfies this requirement of being a

Reviewer 2: Is this really needed? This may make user to do hacks. Imagine we have this interface in the OnStart, they know that the object we return implements both, OnEnd we return only read, then they will guess is the same interface and may do a cast to the ReadWrite interface.

My suggestion is that Span is the write interface and ReadableSpan is the read interface, if access required to both we should pass both interfaces. Maybe I am to "evil" and think of possible hacks and abuses, but I've seen all of these happening in my previous life :)

Author: You could pass a wrapper object that only implements the readable interface to protect against that.

I think we can mention passing the read-only and write-only interface separately as an allowed implementation option for read/write span though.

Reviewer: @Oberon00 suggested I do that hack when I need to mutate the span in `OnEnd` :) I think there is a need to be able to mutate spans globally before export (usually to use other tags as signals to user-defined domain-specific tags) so hope we can find a way. The issue we were worried about is the ordering between `OnEnd` and export

Author: More importantly, spans are already ended at that point and can't be modified anyway. https://github.com/open-telemetry/opentelemetry-specification/pull/669#issuecomment-654841632

Author: @bogdandrutu I incorporated the "read/write implemented as two separate objects" option in 0b9709d6e61bd4b4353a933af30d4be1bd340e5c.

  readable span if it allows to retrieve all information
  that was added to the span (e.g. all attributes, events, (parent) `SpanContext`,
  etc.).
  
  A readable span interface MUST provide a method called `getInstrumentationLibrary` (or similar),
  that returns the `InstrumentationLibrary` information held by the `Tracer` that created the span.
* **Read/write span**: A span interface satisfies this requirement of being a
  read/write span if it provides both the full span API as defined in the
  [API-level definition for span's interface](api.md#span-operations) and
  additionally satisfies the requirements for being a readable span
  as defined in the above bullet point.
  
Note: First, these requirements use "interface" in the most abstract sense --
it does not need to be an actual Java `interface` for example but could also be
a `final` POJO, or, especially in the case of read/write span, even two (or more)
separate objects that are passed together to give full read/write capabilities
(when chosing this implemenation technique,
changes through one object should be reflected in the other
objects immediately if they are observable through them at all

> to avoid introducing suble gotchas).

Reviewer 3: ```suggestion
to avoid introducing subtle gotchas).
```

reference to their `TracerProvider` and access configuration only via this
reference.

## Additional Span interfaces

The [API-level definition for Span's interface](api.md#span-operations)
only defines write-only access to the span.

> This is good because instrumentations and applications are not to use the data

Reviewer 3: ```suggestion
This is good because instrumentations and applications are not meant to use the data
```

stored in a span for application logic.
However, the SDK needs to eventually read back the data in some locations.
Thus, the SDK specification defines two new terms:

* **Readable span**: A span interface that MUST allow to retrieve all information
  that was added to the span (e.g. all attributes, events, (parent) `SpanContext`,

>   etc.).

Reviewer 3: ```suggestion
  etc.).
  The readable span MUST provide a method called `getInstrumentationLibrary` (or similar),
  that returns the `InstrumentationLibrary` information held by the `Tracer` that created the span.
```

or did you intentionally remove this explicit requirement?

Author: Whoops, I meant to move it, but it seems I forgot to paste it. Thanks for catching this!

  A readable span MAY be identical to a read/write span, i.e., SDKs are not
  required to have a read-only interface.

>   SDKs are also allowed to use a read/write span instead of a different readable

Reviewer: ```suggestion
  SDKs are also allowed to use a read/write span instead of a different readable span
```

Author: ```suggestion
  SDKs are also allowed to use a read/write span instead of a different readable interface
```

is what I meant

only defines write-only access to the span.
This is good because instrumentations and applications are not meant to use the data
stored in a span for application logic.
However, the SDK needs to eventually read back the data in some locations.
Thus, the SDK specification defines two new terms:

* **Readable span**: A span interface that MUST allow to retrieve all information
  that was added to the span (e.g. all attributes, events, (parent) `SpanContext`,
  etc.).
  
  The readable span MUST provide a method called `getInstrumentationLibrary` (or similar),
  that returns the `InstrumentationLibrary` information held by the `Tracer` that created the span.
* **Read/write span**: A span that provides both the full span API as defined in the
  [API-level definition for span's interface](api.md#span-operations) and
  additionally the same interface as for readable spans
  as defined in the above bullet point.
  
The readable span interface MAY be identical to the read/write interface,
i.e., SDKs are not required to have a read-only interface.
SDKs are also allowed to use a read/write span instead of a different readable interface

> only in some places where the specification requires a readable span.

Reviewer 2: > SDKs are also allowed to use a read/write span instead of a different readable interface
only in some places where the specification requires a readable span.

This is not clear to me. What are those "some" places? Am I allowed to use everywhere or just in some, if in some what are those "some".

Author: The intention here is to give implementations full flexibility. So when you read the spec and see "readable" in a particular place you might actually get a read/write span. But in another place where the spec says "readable", you may get a read-only span.
I think we currently use readable in the exporter's Export argument and the spanProcessor OnEnd. For example, this allows to have OnEnd receive a read/write span and forward it to the eporter as a read-only span (think SpanData).

Author: After reading your other comment https://github.com/open-telemetry/opentelemetry-specification/pull/669#discussion_r450981708, I think this sentence should exactly resolve the issues you brought up there.

Author: Does c9a5d44fcb880fc559e9c80ee13646f616ec811d help? Or make it even more confusing? I have a hard time coming up with something that is both simple and unambiguous here...

Author: Since no further feedback was given, I'll resolve this.

**Parameters:**

* `Span` - a [read/write span object](#additional-span-interfaces) for the started span.

**Returns:** `Void`

#### OnEnd(Span)

`OnEnd` is called after a span is ended (i.e., the end timestamp is already set).
This method MUST be called synchronously within the [`Span.End()` API](api.md#end),
therefore it should not block or throw an exception.

**Parameters:**

* `Span` - a [readable span object](#additional-span-interfaces) for the ended span.
  Note: Even if the passed Span may be technically writable,

>   since it's already ended at this point, modifying it is not allowed.

Reviewer 3: ```suggestion
  since it's already ended at this point, modifying it is not allowed.
  Furthermore, the order of SpanProcessors being called is not guaranteed and therefore visibility to others could not be relied on anyway.
```

Maybe you could add this clarification to even further discourage use.

Author: That is not true. I'm pretty sure we guarantee that span processors are invoked in the same order they are added.

Reviewer 3: Most likely it will work like this anyway by the way the way it's implemented, better leave it out then.

**Parameters:**

* `Span` - a [readable span object](#additional-span-interfaces) for the started span.
  Note: Since multiple SpanProcessors may be involved, and OnEnd is typically used
  for sending data to exporters, SpanProcessors should rather not modify spans

>   in `OnEnd`, since the resulting behavior would be sensitive to the order of

Reviewer 3: ```suggestion
  in `OnEnd`, even if the passed object would allow this, since the resulting behavior would be sensitive to the order of
```

**Parameters:**

* `Span` - a [readable span object](#additional-span-interfaces) for the started span.
  Note: Since multiple SpanProcessors may be involved, and OnEnd is typically used
  for sending data to exporters, SpanProcessors should rather not modify spans
  in `OnEnd`, even if the passed object would allow this, since the resulting behavior would be sensitive to the order of

>   `SpanProcessors`'s `OnEnd` calls.

Reviewer 4: Sounds great to me, but I'd like also to ask @bogdandrutu for his opinion on this ;)

Author: This Note states a fact and gives a recommendation to implementers of OnEnd. I'd like to leave it here.

I think if we want to better support decoration, we should have two different methods in the interface to invoke when ending a span (or equivalently, have a separate interface for SpanDecorator that has OnDecorate invoked before OnEnd. This could be implemented with a SpanProcessor that is guaranteed to be invoked first at  OnEnd). This is out of scope for this PR I think and 

Your Java spike using ReadWrite span for OnEnd is still conformaning to the spec in this PR though, as it allows substituting a read/write interface in any place the spec requires a readable interface. Alternatively you could document for Java that the ReadableSpan passed to OnEnd may be cast to a ReadWriteSpan. That way, you at least make people read the docs before changing attributes.

Reviewer: Got it - yeah then talking about a more structured approach to decoration in another PR SGTM

#### `Export(batch)`

> Exports a batch of [readable spans](#additional-span-interfaces).

Reviewer 4: I kinda prefer the way this section is at this moment, i.e. `telemetry data`, as it's more concise at what exporters *should* be getting. Or maybe we can say that here a batch of telemetry data _or_ readable spans are exported?

Author: We are in the spec for tracing SDK here, so telemetry data is IMHO needlessly confusing.

> Exports a batch of [readable spans](#additional-span-interfaces).

Reviewer 2: Previously we had 2 "interfaces"/objects:
* ReadableSpan -> allow user to read different parts of the Span or the entire data as "SpanData"
* SpanData -> all data extracted from the Span in an immutable object.

Seems that this PR changes that, I don't know if that was the intention since it is not mentioned in the description.

The transformation from ReadableSpan -> SpanData is important because lazy initialized objects are initialized, and this should happen ideally in only one place and not having every exporter doing that (in case multiple exporters are configured). Also it is important to do it asap to free memory.

Span.End() -> SpanProcessor (uses ReadableSpan to create SpanData) -> SpanExporter (consumes SpanData)

Author: There was no SpanData in the spec previously.

I think the current definition of "readable span" allows both Java's ReadableSpan and SpanData (as per https://github.com/open-telemetry/opentelemetry-specification/pull/669#discussion_r450996960, it also allows implementing a readable span with SpanData in some places and ReadableSpan in others).

Author: @bogdandrutu ping!

Reviewer 4: Ping @bogdandrutu We need your feedback this PR gets either merged or iterated on ;)

Author: Since no further feedback was given, I'll resolve this.

**Parameters:**

> batch - a batch of telemetry data. The exact data type of the batch is language

Reviewer 3: ```suggestion
batch - a batch of readable spans. The exact data type of the batch is language
```
as defined above

Author: One moment, this needs to be linked.

specific, typically it is a list of [readable spans](#additional-span-interfaces),

> e.g. for spans in Java it will be typically `Collection<ExportableSpan>`.

Reviewer 3: ```suggestion
e.g. for spans in Java it will be typically `Collection<SpanData>`.
```
